### PR TITLE
[python-package] limit when num_boost_round warnings are emitted (fixes #6324)

### LINKED ISSUE #6324: [python-package] UserWarning: Found `...` in params
## Description

This and many other similar issues have been reported along the years. Yet I cannot find a reasonable solution. lightgbm is constantly throwing uninteresting warnings about aliases. They are not even easy to locally suppress once you start running subprocesses (n_jobs = -1 or > 1) and have to deal with PYTHONWARNINGS in the environment. In such cases there is a long stream of warnings, coming for example from a grid search. The verbose argument does little to silence lightgbm since the warnings are thrown at Python side.

[It was suggested](https://github.com/microsoft/LightGBM/issues/3676#issuecomment-765391073) to use a parameter's name instead of its alias, but take for example `num_iterations` that is documented in https://lightgbm.readthedocs.io/en/latest/Parameters.html as a name, not an alias: it nevertheless throws a warning and you have to try every other alias in order to discover that `n_iterations` is the right one.

I think some action has to be taken here.

## Reproducible example

```python
import numpy as np
import pandas as pd
import lightgbm as lgb

X = pd.DataFrame(dict(x1=np.linspace(0, 1, 100), x2=np.linspace(0, 1, 100)))
y = 2 * X.x1 + 3 * X.x2
lgb.LGBMRegressor(verbose=-1, num_iterations=10).fit(X, y)
```

## Environment info

LightGBM version or commit hash: 4.3.0

Command(s) you used to install LightGBM: 

```shell
pip install lightgbm
```

macOS Sonoma
Python 3.11.7

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #6324

Replaces #6548

* ensures that the warning about aliases for `num_iterations` is only raised if the user-provided configuration has conflicting values
* expands documentation about how parameters in LightGBM are merged

## Notes for Reviewers

Just opening this to show I'm working on it. It still needs tests, and isn't ready for review yet.

### How I tested this

Enabled this branch on readthedocs. See https://readthedocs.org/projects/lightgbm/builds/.

### Isn't this kind of a lot of complexity for one parameter?

I think it's worth it. "number of boosting rounds" is one of the most important parameters for gradient boosting. I think it's worth handling it very carefully and raising an informative warning in the cases where conflicting configuration is provided. For example, I think this warning could save people from unnecessarily performing hyperparameter tuning over different combinations of `num_iterations` and `num_trees`, not understanding that those mean the same thing in LightGBM's parameters.

### Why not use `pytest.parametrize()` for all these tests?

I started out that way, but then decided that it'd require enough if-else blocks and other logic that I was worried about accidentally, silently, mishandling some test cases.

I think the level of repetition in these test cases is ok, because it makes them easy to read and understand.

### CODE REVIEW

File: tests/python_package_test/utils.py [hunk 1]
```diff
@@ -191,6 +191,25 @@ def pickle_and_unpickle_object(obj, serializer):
     return obj_from_disk  # noqa: RET504
 
 
+def assert_silent(capsys) -> None:
+    """
+    Given a ``CaptureFixture`` instance (from the ``pytest`` built-in ``capsys`` fixture),
+    read the recently-captured data into a variable and assert that nothing was written
+    to stdout or stderr.
+
+    This is just hear to turn 3 lines of boilerplate into 1.
```
Comment (jameslamb) [inline tests/python_package_test/utils.py:200]:
```suggestion
    This is just here to turn 3 lines of repetitive code into 1.
```

File: python-package/lightgbm/engine.py [hunk 2]
```diff
@@ -198,11 +254,10 @@ def train(
...
         params["objective"] = "none"
-    for alias in _ConfigAliases.get("num_iterations"):
-        if alias in params:
-            num_boost_round = params.pop(alias)
-            _log_warning(f"Found `{alias}` in params. Will use it instead of argument")
-    params["num_iterations"] = num_boost_round
+
+    params = _choose_num_iterations(num_boost_round_kwarg=num_boost_round, params=params)
+    num_boost_round = params["num_iterations"]
+
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:260]: I think the following line should be moved here now
https://github.com/microsoft/LightGBM/blob/63e15300ec88c143f57e1214d10e3d6aa87b92c4/python-package/lightgbm/engine.py#L228-L229

File: python-package/lightgbm/engine.py [hunk 3]
```diff
@@ -733,11 +788,10 @@ def cv(
...
         params["objective"] = "none"
-    for alias in _ConfigAliases.get("num_iterations"):
-        if alias in params:
-            _log_warning(f"Found '{alias}' in params. Will use it instead of 'num_boost_round' argument")
-            num_boost_round = params.pop(alias)
-    params["num_iterations"] = num_boost_round
+
+    params = _choose_num_iterations(num_boost_round_kwarg=num_boost_round, params=params)
+    num_boost_round = params["num_iterations"]
+
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:794]: I think the following line should be moved here now
https://github.com/microsoft/LightGBM/blob/63e15300ec88c143f57e1214d10e3d6aa87b92c4/python-package/lightgbm/engine.py#L771-L772

File: python-package/lightgbm/engine.py [hunk 4]
```diff
@@ -733,11 +788,10 @@ def cv(
...
         fobj = params["objective"]
         params["objective"] = "none"
-    for alias in _ConfigAliases.get("num_iterations"):
-        if alias in params:
-            _log_warning(f"Found '{alias}' in params. Will use it instead of 'num_boost_round' argument")
-            num_boost_round = params.pop(alias)
-    params["num_iterations"] = num_boost_round
+
+    params = _choose_num_iterations(num_boost_round_kwarg=num_boost_round, params=params)
+    num_boost_round = params["num_iterations"]
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:793]: Is it OK to leave `num_iterations` key in params and pass them to cpp side, instead of pop-ing it like before?

File: tests/python_package_test/test_sklearn.py [hunk 5]
```diff
@@ -1336,6 +1337,51 @@ def test_verbosity_is_respected_when_using_custom_objective(capsys):
...
+    reg = lgb.LGBMRegressor(**base_kwargs, n_iter=3, num_iterations=3).fit(X, y)
+    assert reg.n_estimators_ == 3
+    assert_silent(capsys)
+
+    # no warning: 4 aliases (all same value)
+    reg = lgb.LGBMRegressor(**base_kwargs, n_iter=3, num_trees=3, nrounds=3, max_iter=3).fit(X, y)
+    assert reg.n_estimators_ == 3
+    assert_silent(capsys)
+
+    # warning: 2 aliases (different values... first one in the order from Config::parameter2aliases() wins)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:1377]: Please add one more test case from `test_engine`:
> \# warning: 2 aliases (different values... "num_iterations" wins because it's the main param name)

File: tests/python_package_test/utils.py [hunk 6]
```diff
@@ -191,6 +191,25 @@ def pickle_and_unpickle_object(obj, serializer):
...
+
+    Note that this does have a side effect... ``capsys.readouterr()`` copies
+    from a buffer then frees it. So it will only store into ``.out`` and ``.err`` the
+    captured output since the last time that ``.readouterr()`` was called.
+
+    ref: https://docs.pytest.org/en/stable/how-to/capture-stdout-stderr.html
+    """
+    captured = capsys.readouterr()
+    assert captured.out == ""
+    assert captured.err == ""
```
Comment (StrikerRUS) [inline tests/python_package_test/utils.py:210]: In case of failure, to see in logs what actual value was captured
```suggestion
    assert captured.out == "", captured.out
    assert captured.err == "", captured.err
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
I like the general idea of this PR!
Left some minor comments below:

File: docs/Parameters.rst [hunk 7]
```diff
@@ -17,18 +17,61 @@ This page contains descriptions of all parameters in LightGBM.
 Parameters Format
 -----------------
 
+Parameters are merged together in the following order (later items overwrite earlier ones):
+
+1. LightGBM's default values
+2. (CLI only) configuration in a file passed like ``config=train.conf``
+3. (CLI only) configuration passed via the command line
+4. (Python, R) ``params`` function argument
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply docs/Parameters.rst:25]: I agree with both of these, thanks! Had forgotten about those extra files. I'll added that here.

File: python-package/lightgbm/engine.py (hunk 2 above)
Comment (jameslamb) [reply python-package/lightgbm/engine.py:260]: Agreed! I think maybe it always should have been after this processing, to prevent **any** values `<= 0` from making it into `"num_iterations"` being passed to the C++ side.

File: python-package/lightgbm/engine.py (hunk 4 above)
Comment (jameslamb) [reply python-package/lightgbm/engine.py:793]: It was already being passed to the cpp side... the result of all the processing that resolves keyword arguments and `params` entries is assigned back to `params["num_iterations"]` right here: https://github.com/microsoft/LightGBM/blob/d67ecf94259217746b031a945d66f4e0fcec8579/python-package/lightgbm/engine.py#L205

So I think this PR does not change the behavior of what is passed through to the C++ side.

Comment (StrikerRUS) [reply python-package/lightgbm/engine.py:793]: Ah, indeed! Overlooked this moment, sorry.

File: tests/python_package_test/utils.py (hunk 6 above)
Comment (jameslamb) [reply tests/python_package_test/utils.py:210]: Good suggestion, agreed!

File: docs/Parameters.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply docs/Parameters.rst]: updated in https://github.com/microsoft/LightGBM/pull/6579/commits/b6b5c806c840640b348fe373aa4255fd83bce4b5

[the remark this answers is not in the document]
Comment (jameslamb) [reply docs/Parameters.rst]: updated in https://github.com/microsoft/LightGBM/pull/6579/commits/b6b5c806c840640b348fe373aa4255fd83bce4b5. I proposed slightly different language there, let me know what you think.

File: python-package/lightgbm/engine.py (hunk 2 above)
Comment (jameslamb) [reply python-package/lightgbm/engine.py:259]: updated in https://github.com/microsoft/LightGBM/pull/6579/commits/b6b5c806c840640b348fe373aa4255fd83bce4b5

File: python-package/lightgbm/engine.py (hunk 3 above)
Comment (jameslamb) [reply python-package/lightgbm/engine.py:792]: updated in https://github.com/microsoft/LightGBM/pull/6579/commits/b6b5c806c840640b348fe373aa4255fd83bce4b5

File: tests/python_package_test/test_sklearn.py (hunk 5 above)
Comment (jameslamb) [reply tests/python_package_test/test_sklearn.py:1387]: ah yes you're totally right, added in https://github.com/microsoft/LightGBM/pull/6579/commits/b6b5c806c840640b348fe373aa4255fd83bce4b5

File: docs/Parameters.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply docs/Parameters.rst]: Great!

File: docs/Parameters.rst [hunk 8]
```diff
@@ -17,18 +17,63 @@ This page contains descriptions of all parameters in LightGBM.
 Parameters Format
 -----------------
 
+Parameters are merged together in the following order (later items overwrite earlier ones):
+
+1. LightGBM's default values
+2. (CLI only) configuration in a file passed like ``config=train.conf``
+3. special files for ``weight``, ``init_score``, ``query``, and ``positions`` (see `Others <#others>`__)
```
Comment (StrikerRUS) [inline docs/Parameters.rst:24]: I believe 2 and 3 should be switched in the order.

For example, according to https://github.com/microsoft/LightGBM/blob/master/docs/Parameters.rst#weight-data
> LightGBM will load the weight file automatically if it exists.
> Also, you can include weight column in your data file.

So, `weight_column` will overwrite `train.txt.weight` file.

File: python-package/lightgbm/engine.py [hunk 9]
```diff
@@ -733,11 +784,12 @@ def cv(
...
+    if num_boost_round <= 0:
+        raise ValueError(f"Number of boosting rounds must be greater than 0. Got {num_boost_round}.")
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:791]: Thanks for changing `num_boost_round` to `Number of boosting rounds`.
👍

File: tests/python_package_test/test_sklearn.py [hunk 10]
```diff
@@ -1337,6 +1338,60 @@ def test_verbosity_is_respected_when_using_custom_objective(capsys):
...
+    assert_silent(capsys)
+
+    # no warning: 4 aliases (all same value)
+    reg = lgb.LGBMRegressor(**base_kwargs, n_iter=3, num_trees=3, nrounds=3, max_iter=3).fit(X, y)
+    assert reg.n_estimators_ == 3
+    assert_silent(capsys)
+
+    # warning: 2 aliases (different values... "num_iterations" wins because it's the main param name)
+    # with pytest.warns(UserWarning, match="LightGBM will perform up to 5 boosting rounds"):
+    #     reg = lgb.LGBMRegressor(**base_kwargs, num_iterations=-8, n_iter=6).fit(X, y)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:1380]: Please remove these commented lines.

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
Thanks a lot for addressing my comments!
I have two more suggestions based on the most recent commit.

File: python-package/lightgbm/engine.py (hunk 9 above)
Comment (jameslamb) [reply python-package/lightgbm/engine.py:791]: Yep definitely! Hopefully this will help reduce confusion for some users.

File: tests/python_package_test/test_sklearn.py (hunk 10 above)
Comment (jameslamb) [reply tests/python_package_test/test_sklearn.py]: 😅 ah sorry, left this in from testing. Thanks as always for your attention to detail! Removed in https://github.com/microsoft/LightGBM/pull/6579/commits/e5de60ab0674eeb1ac0e4dc03f2b201d624d2d79

File: docs/Parameters.rst (hunk 8 above)
Comment (jameslamb) [reply docs/Parameters.rst]: ohhhhh `weight_column`, right!

I agree they should be switched in order. Did that in https://github.com/microsoft/LightGBM/pull/6579/commits/e5de60ab0674eeb1ac0e4dc03f2b201d624d2d79.

**Review by StrikerRUS — APPROVED** [review]
Thanks very much for working on this change!

**Comment by jameslamb** [comment]
Thanks for the thoughtful reviews @StrikerRUS !

**Comment by StrikerRUS** [comment]
@jameslamb Could you please remove [python/alias-warning](https://lightgbm.readthedocs.io/en/python-alias-warning/) branch on RTD?

I think we don't need to leave debug builds there.

image

**Comment by jameslamb** [comment]
🙈 sorry about that, thanks for your attention to detail!

Just removed it.
